### ISSUE TITLE
ob1/vader freezes when built with older gcc

### ISSUE THREAD

**Issue by tvegas1** [open-mpi/ompi#11772]
## Problem
Following command freezes (busy loop) on v4.1.x (9a9ddfc583923dea91ed8913f5188dfbf560f2c0):
```
mpirun -mca coll '^hcoll' -mca pml ob1 -mca btl vader,self -mca btl_vader_single_copy_mechanism xpmem -np 2 osu_alltoall
# OSU MPI All-to-All Personalized Exchange Latency Test v5.8
# Size       Avg Latency(us)
```

## Analysis/Question
Reverting d9895ce0de1abd488db5ce3debc4f9b5e77d599f fixes the issue. In particular the two macros below are missing:
- `PLATFORM_COMPILER_GNU`
  - Seems to be coming from `opal/include/opal/opal_portable_platform.h`
- `PLATFORM_ARCH_X86_64`
  - Seems to be coming from openpmix

Is there something missing in my setup/configure or shall the backport d9895ce0de1abd488db5ce3debc4f9b5e77d599f  rather be only touching the `__GNUC__` version number?

If the issue is not on my side, that could be problematic for other use cases, when building with gcc < 8.

## Build/Configure
gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC)
From clean git at 9a9ddfc583923dea91ed8913f5188dfbf560f2c0:
```
./autogen.pl
./configure \
    --prefix=`pwd`/rfs \
    --with-ucx=`pwd`/../ucx/rfs \
    --with-xpmem=`pwd`/../../../xpmem/rfs \
    --with-xpmem-libdir=`pwd`/../../../xpmem/rfs/lib \
    --with-libevent=internal \
    --enable-mpi1-compatibility \
    --with-slurm \
    --with-platform=contrib/platform/mellanox/optimized \
    --disable-man-pages
make -j 28 && make install
```

**Comment by ggouaillardet** [comment]
Thanks for the detailed report!

@devreal can you please have a look?

It seems this should have been a backport instead of a cherry-pick.

For example `PLATFORM_ARCH_X86_64` is defined in the `main` branch in `opal/include/opal/opal_portable_platform_real.h` but there is no such file in the `v4.1` series.

**Comment by devreal** [comment]
Thanks for the report @tvegas1. Just to confirm: does the following patch fix the problem?

```diff
diff --git a/opal/include/opal/sys/gcc_builtin/atomic.h b/opal/include/opal/sys/gcc_builtin/atomic.h
index e016ca695e..10bd34309d 100644
--- a/opal/include/opal/sys/gcc_builtin/atomic.h
+++ b/opal/include/opal/sys/gcc_builtin/atomic.h
@@ -52,7 +52,7 @@
 #define OPAL_HAVE_ATOMIC_SWAP_64 1
 
 
-#if defined(PLATFORM_ARCH_X86_64) && defined(PLATFORM_COMPILER_GNU) && __GNUC__ < 8
+#if (OPAL_ASSEMBLY_ARCH == OPAL_X86_64) && defined (__GNUC__) && !defined(__llvm) && __GNUC__ < 8
     /* work around a bug in older gcc versions where ACQUIRE seems to get
      * treated as a no-op instead */
 #define OPAL_BUSTED_ATOMIC_MB 1
```

Thanks @ggouaillardet for the hint, I was a bit puzzled but it makes sense that the cherry-pick was wrong.

**Comment by tvegas1** [comment]
Tested that the issue is fixed with that patch.

**Comment by jsquyres** [comment]
Fixed by #11773.
